analyzers/src/SonarAnalyzer.Common/AnalysisContext/SonarAnalysisContextBase.cs

```diff
@@ -131,11 +131,21 @@ public bool HasMatchingScope(DiagnosticDescriptor descriptor)
             descriptor.CustomTags.Contains(tag);
     }
 
-    private bool IsExcluded(SonarLintXmlReader sonarLintXml, string filePath) =>
+    private bool IsExcluded(SonarLintXmlReader sonarLintXml, string filePath)
+    {
         // If ProjectType is not 'Unknown' it means we are in S4NET context and all files are analyzed.
         // If ProjectType is 'Unknown' then we are in SonarLint or NuGet context and we need to check if the file has been excluded from analysis through SonarLint.xml.
-        ProjectConfiguration().ProjectType == ProjectType.Unknown
-        && !FileInclusionCache.GetValue(Compilation, _ => new()).GetOrAdd(filePath, _ => sonarLintXml.IsFileIncluded(filePath, IsTestProject()));
+        if (ProjectConfiguration().ProjectType == ProjectType.Unknown)
+        {
+            var fileInclusionCache = FileInclusionCache.GetOrCreateValue(Compilation);
+            // Hotpath: Don't use GetOrAdd with the value factory parameter. It allocates a delegate which causes GC preasure.
+            var isIncluded = fileInclusionCache.TryGetValue(filePath, out var result)
+                ? result
+                : fileInclusionCache.GetOrAdd(filePath, sonarLintXml.IsFileIncluded(filePath, IsTestProject()));
+            return !isIncluded;
```

Comment:
This return doesn't seem right. It should return the result of the "new value", that is a result of the newly added`sonarLintXml.IsFileIncluded(filePath, IsTestProject())`

Reply:
FileInclusionCache is a ConcurrentDictionary. The return value is the newly created value (or from the cache if it is already there by the time we get there). It is a bit misleading here because `TValue` is a bool.

[ConcurrentDictionary<TKey,TValue>.GetOrAdd Method](https://learn.microsoft.com/en-us/dotnet/api/system.collections.concurrent.concurrentdictionary-2.getoradd?view=net-7.0#system-collections-concurrent-concurrentdictionary-2-getoradd(-0-1))

> Returns
TValue
The value for the key. This will be either the existing value for the key if the key is already in the dictionary, or the new value if the key was not in the dictionary.